### PR TITLE
add graf zahl (spark + kafka example)

### PR DESCRIPTION
**Pull request by mattf**

### CODE REVIEW

**Comment by mattf** [comment]
@elmiko @willb ptal

File: _applications/grafzahl.md [hunk 1]
```diff
@@ -0,0 +1,120 @@
...
+- https://github.com/mattf/openshift-kafka
+---
+
+<h1 id="introduction">Introduction</h1>
+
+This is intended to be an example of how an application can process
+data from Apache Kafka with Apache Spark on OpenShift.
+
+Graf Zahl will count, as he does, the words on an Apache Kafka topic
+and display the top-k on in a web page for the user. There isn't much
```
Comment (elmiko) [inline _applications/grafzahl.md]: "top-k on in a web page", i think there is one too many words or one too few

File: _applications/grafzahl.md [hunk 2]
```diff
@@ -0,0 +1,120 @@
...
+oc new-app openshift/python-27-centos7~https://github.com/mattf/word-fountain -e SERVERS=apache-kafka:9092
+```
+
+Forth, launch Graf Zahl himself, using the Oshinko pyspark S2I
+builder. This command uses `-p` parameters and `-e` environment
+variables. The `oshinko-pyspark-build-dc` template uses the
+parameters, while the [grafzahl
+app.py](https://github.com/mattf/grafzahl/blob/master/app.py) uses the
+environment variables.
+
```
Comment (elmiko) [inline _applications/grafzahl.md]: i think we should have a note in here to let readers know that there will be some associated build time with the word-fountain and grafzahl images, just so they know to check the builds and what not.

**Comment by elmiko** [comment]
one other thing i just noticed, i ran the command from the end of your asciinema and it did not return a json object, instead i got the html from the webserver

```
$ curl $(oc get route/grafzahl --template 'http://{{.spec.host}}')/
<html>
<body>
<link rel="stylesheet" href="http://www.patternfly.org/assets/css/patternfly-site.min.css"></link>
<link rel="stylesheet" href="http://www.patternfly.org/assets/css/patternfly-site.min.css"></link>
<link rel="stylesheet" href="http://www.patternfly.org/assets/css/patternfly-site.min.css"></link>
<script src="http://www.patternfly.org/components/jquery/dist/jquery.min.js"></script>
<script src="http://www.patternfly.org/components/bootstrap/dist/js/bootstrap.min.js"></script>
<script src="http://www.patternfly.org/components/patternfly/dist/js/patternfly.min.js"></script>
<script src="http://www.patternfly.org/components/c3/c3.min.js"></script>
<script src="http://www.patternfly.org/components/d3/d3.min.js"></script>
<div id="bar-chart"></div>
...
```

**Comment by mattf** [comment]
full command at the end of the asciicast is: curl $(oc get route/grafzahl --template 'http://{{.spec.host}}')/data

File: _applications/grafzahl.md (hunk 1 above)
Comment (mattf) [inline _applications/grafzahl.md]: fixing

File: _applications/grafzahl.md (hunk 2 above)
Comment (mattf) [inline _applications/grafzahl.md]: fyi steps 3 & 4 can happen in parallel or opposite order. there's just nothing to look at until the word-fountain starts up.

**Comment by elmiko** [comment]
ok, using `/data` did work for me, although my grafzahl is still not receiving data from the word fountain. how could i debug this?

**Comment by mattf** [comment]
in the word-fountain logs you should see something from https://github.com/mattf/word-fountain/blob/master/app.py#L21

https://github.com/mattf/word-fountain/blob/master/app.py#L33 should can be enabled if you use your own word-fountain. i should add a --debug...

**Comment by elmiko** [comment]
the logs for my word-fountain only show this:

```
$ oc logs word-fountain-1-9kth9
---> Running application from Python script (app.py) ...
```

after running a kafka-debug pod, i can see that words are being placed on the topic, but grafvahl doesn't seem to be getting them.

**Comment by mattf** [comment]
did oshinko spin up a spark cluster for grafzahl?

are there any errors in the grafzahl log?

**Comment by elmiko** [comment]
a cluster is spawned, and the grafzahl logs don't show anything looks suspicious. after all the setup and connection to the spark cluster, i just see this:

```
172.17.0.1 - - [24/Feb/2017 16:46:45] "GET /data?n=10 HTTP/1.1" 200 -
INFO:werkzeug:172.17.0.1 - - [24/Feb/2017 16:46:45] "GET /data?n=10 HTTP/1.1" 200 -
17/02/24 16:46:55 INFO SparkSqlParser: Parsing command: SELECT * FROM results ORDER BY count DESC LIMIT 10
17/02/24 16:46:55 INFO SparkContext: Starting job: collect at /opt/app-root/src/app.py:18
17/02/24 16:46:55 INFO DAGScheduler: Got job 18 (collect at /opt/app-root/src/app.py:18) with 1 output partitions
17/02/24 16:46:55 INFO DAGScheduler: Final stage: ResultStage 18 (collect at /opt/app-root/src/app.py:18)
17/02/24 16:46:55 INFO DAGScheduler: Parents of final stage: List()
17/02/24 16:46:55 INFO DAGScheduler: Missing parents: List()
17/02/24 16:46:55 INFO DAGScheduler: Submitting ResultStage 18 (MapPartitionsRDD[76] at collect at /opt/app-root/src/app.py:18), which has no missing parents
17/02/24 16:46:55 INFO MemoryStore: Block broadcast_18 stored as values in memory (estimated size 3.9 KB, free 366.2 MB)
17/02/24 16:46:55 INFO MemoryStore: Block broadcast_18_piece0 stored as bytes in memory (estimated size 2.3 KB, free 366.2 MB)
17/02/24 16:46:55 INFO BlockManagerInfo: Added broadcast_18_piece0 in memory on 172.17.0.14:33639 (size: 2.3 KB, free: 366.3 MB)
17/02/24 16:46:55 INFO SparkContext: Created broadcast 18 from broadcast at DAGScheduler.scala:996
17/02/24 16:46:55 INFO DAGScheduler: Submitting 1 missing tasks from ResultStage 18 (MapPartitionsRDD[76] at collect at /opt/app-root/src/app.py:18)
17/02/24 16:46:55 INFO TaskSchedulerImpl: Adding task set 18.0 with 1 tasks
17/02/24 16:46:55 INFO TaskSetManager: Starting task 0.0 in stage 18.0 (TID 18, 172.17.0.11, executor 0, partition 0, PROCESS_LOCAL, 7759 bytes)
17/02/24 16:46:55 INFO BlockManagerInfo: Added broadcast_18_piece0 in memory on 172.17.0.11:33555 (size: 2.3 KB, free: 366.3 MB)
17/02/24 16:46:55 INFO TaskSetManager: Finished task 0.0 in stage 18.0 (TID 18) in 14 ms on 172.17.0.11 (executor 0) (1/1)
17/02/24 16:46:55 INFO TaskSchedulerImpl: Removed TaskSet 18.0, whose tasks have all completed, from pool 
17/02/24 16:46:55 INFO DAGScheduler: ResultStage 18 (collect at /opt/app-root/src/app.py:18) finished in 0.015 s
17/02/24 16:46:55 INFO DAGScheduler: Job 18 finished: collect at /opt/app-root/src/app.py:18, took 0.017843 s
```

and it's repeated for each GET

**Comment by mattf** [comment]
so kafka is up, word-fountain is up, word-fountain is producing, grafzahl's flask app is up and running, the spark cluster is up, only thing left is the spark streaming part of grafzahl.

if you go to the spark master web ui is there an app connected?
do you see grafzahl periodically kicking off spark jobs not associated w/ a GET?

**Comment by elmiko** [comment]
yes, the spark master ui shows an app connected.

i closed the browser connection to grafzahl to quite the GETs and after a few minutes i see no activity from it, i'm guessing it's just not connecting to the kafka topic or something?

**Comment by mattf** [comment]
maybe, try to redeploy the grafzahl dc

**Comment by elmiko** [comment]
tried redeploying already, didn't work.

but, i just checked the grafzahl pod and i see that the `SERVERS` env is not getting propagated, and i noticed this in the `oc new-app` help:

```
-e, --env=[]: Specify a key-value pair for an environment variable to set into each container. This doesn't apply to
objects created from a template, use parameters instead.
```

i think the `-e` is not getting passed because we are using the template form of instantiation

**Comment by mattf** [comment]
what's in the env for your grafzahl pod?

```
$ oc describe po/pyspark-vjdl-1-yl1se | grep -A8 Environment
    Environment Variables:
      SERVERS:				apache-kafka:9092
      OSHINKO_CLUSTER_NAME:		
      APP_ARGS:				
      SPARK_OPTIONS:			--packages org.apache.spark:spark-sql-kafka-0-10_2.11:2.1.0
      OSHINKO_DEL_CLUSTER:		true
      APP_EXIT:				false
      OSHINKO_NAMED_CONFIG:		
      OSHINKO_SPARK_DRIVER_CONFIG:	
```

**Comment by elmiko** [comment]
```
$ oc describe po/pyspark-ve81-1-r36gt | grep -A8 Environment
    Environment Variables:
      OSHINKO_CLUSTER_NAME:
      APP_ARGS:
      SPARK_OPTIONS:                    --packages org.apache.spark:spark-sql-kafka-0-10_2.11:2.1.0
      OSHINKO_DEL_CLUSTER:              true
      APP_EXIT:                         false
      OSHINKO_NAMED_CONFIG:
      OSHINKO_SPARK_DRIVER_CONFIG:
Conditions:
```

**Comment by mattf** [comment]
also, look at your word-fountain pod, it uses SERVERS env too and you reported it as working.

**Comment by elmiko** [comment]
ok, so i got it working with the 1.5 version of openshift origin. looks like this is an issue with 1.4.

**Comment by mattf** [comment]
i've reproduced the issue using v1.4.1+3f9807a

the SERVERS env is indeed not passed through

**Comment by elmiko** [comment]
thanks Matt!

**Pull request merged**